RavenDB-24102 & RavenDB-24103 & RavenDB-23030 Querying async optimizations

maciejaszyk opened the pull request:
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-23030
https://issues.hibernatingrhinos.com/issue/RavenDB-24102 
https://issues.hibernatingrhinos.com/issue/RavenDB-24103 
### Additional description

This PR focuses on reducing the overhead of the async state machine in the queries handler. It removes unnecessary awaits/async, creates fast-sync paths for when the code is executed synchronously, and refactors the handler code to be smaller for some paths.

The optimized codepath is common for all tickets linked above.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-24102 Removes await/async from places that don't need it.

maciejaszyk committed:
RavenDB-24102 QueryHandler async optimizations.

maciejaszyk committed:
RavenDB-24102 Introducing Requests.Benchmark

maciejaszyk committed:
RavenDB-24102 Simplifying the blittable text writer & Benchmark fixes

arekpalinski commented on bench/Voron.Benchmark/Voron.Benchmark.csproj:
```diff
@@ -17,6 +17,7 @@
     <PackageReference Include="BenchmarkDotNet" />
     <PackageReference Include="Microsoft.CSharp" />
     <PackageReference Include="Microsoft.Diagnostics.Tracing.TraceEvent" />
+    <PackageReference Include="Microsoft.Diagnostics.Tracing.TraceEvent"  />
```
Duplicated reference added

maciejaszyk replied:
Removed

arekpalinski reviewed (commented):
I talked with @maciejaszyk offline about this PR. There will be some adjustments to the PR. Please hold on with the review.

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessor.cs:
```diff
@@ -41,33 +41,38 @@ protected RequestTimeTracker CreateRequestTimeTracker()
+            : GetIndexQueryFromPostRequestAsync();
 
-        var json = await context.ReadForMemoryAsync(_stream, "index/query");
+        async ValueTask<IndexQueryServerSide> GetIndexQueryFromPostRequestAsync()
```
Is it context capturing? If yes, can we make it static?

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -37,32 +38,29 @@ protected AbstractQueriesHandlerProcessorForGet([NotNull] TRequestHandler reques
         {
-            await ServerSideQueryAsync(context, query);
-            return;
+            return ServerSideQueryAsync(context, query);
```
You `return VT`. The last line of this method should properly wrap the exception as a `ValueTask` as well.

maciejaszyk replied:
Done

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -37,32 +38,29 @@ protected AbstractQueriesHandlerProcessorForGet([NotNull] TRequestHandler reques
     }
 
-    protected abstract ValueTask<IndexEntriesQueryResult> GetIndexEntriesAsync(TQueryContext queryContext, TOperationContext context, IndexQueryServerSide query, long? existingResultEtag, bool ignoreLimit, OperationCancelToken token);
+    protected abstract Task<IndexEntriesQueryResult> GetIndexEntriesAsync(TQueryContext queryContext, TOperationContext context, IndexQueryServerSide query, long? existingResultEtag, bool ignoreLimit, OperationCancelToken token);
```
Why the change here?

maciejaszyk replied:
Answer here: https://github.com/ravendb/ravendb/pull/21142/files#r2414177056

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -78,24 +76,27 @@ private async ValueTask IndexEntriesAsync(TQueryContext queryContext, TOperation
         {
-            await writer.WriteIndexEntriesQueryResultAsync(context, result, token.Token);
+            var writeIndexEntriesQueryResultsTask = writer.WriteIndexEntriesQueryResultAsync(context, result, token.Token);
+            if (writeIndexEntriesQueryResultsTask.IsCompletedSuccessfully == false)
```
I don't think it matters. I believe the check is emitted anyway for you.

maciejaszyk replied:
Well, if not, we will get exactly what we want. Otherwise, nothing changes, right?

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -78,24 +76,27 @@ private async ValueTask IndexEntriesAsync(TQueryContext queryContext, TOperation
     protected abstract ValueTask ExplainAsync(TQueryContext queryContext, IndexQueryServerSide query, OperationCancelToken token);
 
-    protected abstract ValueTask<FacetedQueryResult> GetFacetedQueryResultAsync(IndexQueryServerSide query, TQueryContext queryContext, long? existingResultEtag, OperationCancelToken token);
+    protected abstract Task<FacetedQueryResult> GetFacetedQueryResultAsync(IndexQueryServerSide query, TQueryContext queryContext, long? existingResultEtag, OperationCancelToken token);
```
What is the reason to go with `Task`?

maciejaszyk replied:
Answer here: https://github.com/ravendb/ravendb/pull/21142/files#r2414177056

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -78,24 +76,27 @@ private async ValueTask IndexEntriesAsync(TQueryContext queryContext, TOperation
-    public override async ValueTask ExecuteAsync()
+    public async Task ExecuteAsTask()
     {
+        using (this)
```
If we had a registration for disposables, do you think we could have all these using eliminated?

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -104,7 +105,8 @@ public override async ValueTask ExecuteAsync()
                     var parameters = QueryStringParameters.Create(HttpContext.Request);
-                    var indexQuery = await GetIndexQueryAsync(context, QueryMethod, tracker, parameters.AddSpatialProperties).AsTask();
+                    var indexQueryTask = GetIndexQueryAsync(context, QueryMethod, tracker, parameters.AddSpatialProperties);
+                    var indexQuery = indexQueryTask.IsCompletedSuccessfully ? indexQueryTask.Result : await indexQueryTask;
```
I'd check but my assumption is that you'll get this generated for free by the compiler.

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -119,77 +121,9 @@ public override async ValueTask ExecuteAsync()
-                        AddQueryTimingsToTrafficWatch(indexQuery);
-                    }
+                    var processQueryTask = ProcessQueryAsync(parameters, indexQuery, queryContext, context, existingResultEtag, token);
+                    if (processQueryTask.IsCompletedSuccessfully == false)
```
Similar to the above. I assume this will be generated automatically by the compiler.

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -275,7 +297,10 @@ private async ValueTask HandleSuggestQueryAsync(IndexQueryServerSide query, TQue
-            (numberOfResults, totalDocumentsSizeInBytes) = await writer.WriteSuggestionQueryResultAsync(operationContext, result, token.Token);
+            var writeSuggestionQueryResultTask = writer.WriteSuggestionQueryResultAsync(operationContext, result, token.Token);
+            (numberOfResults, totalDocumentsSizeInBytes) = writeSuggestionQueryResultTask.IsCompletedSuccessfully 
+                ? writeSuggestionQueryResultTask.Result 
```
Similar to the above. I'd check if it's not done by the compiler

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/DatabaseQueriesHandlerProcessorForGet.cs:
```diff
@@ -34,32 +34,32 @@ protected override IDisposable AllocateContextForQueryOperation(out QueryOperati
     protected override RavenConfiguration Configuration => RequestHandler.Database.Configuration;
 
-    protected override async ValueTask<FacetedQueryResult> GetFacetedQueryResultAsync(IndexQueryServerSide query, QueryOperationContext queryContext,
+    protected override Task<FacetedQueryResult> GetFacetedQueryResultAsync(IndexQueryServerSide query, QueryOperationContext queryContext,
```
Why?

maciejaszyk replied:
The QueryRunner is a Task, previously we had a pattern `return await QueryRunner.Method()` to match the VT type. This way, we can just return actual task without additional async method.

Scooletz commented on src/Raven.Server/Documents/Handlers/QueriesHandler.cs:
```diff
@@ -8,17 +8,15 @@ namespace Raven.Server.Documents.Handlers
         {
-            using (var processor = new DatabaseQueriesHandlerProcessorForGet(this, HttpMethod.Post))
-                await processor.ExecuteAsync();
+            return new DatabaseQueriesHandlerProcessorForGet(this, HttpMethod.Post).ExecuteAsTask();
```
So this is eliding.. Ok

Scooletz commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -450,8 +450,11 @@ public static void WriteExplanation(this AbstractBlittableJsonTextWriter writer,
-
+            var numberOfResultsTask = writer.WriteQueryResultAsync(context, result, metadataOnly: false, partial: true, token);
+            var numberOfResults = numberOfResultsTask.IsCompletedSuccessfully 
+                ? numberOfResultsTask.Result 
```
Similar to the above. Isn't it implemented by the compiler?

maciejaszyk replied:
Answer here: https://github.com/ravendb/ravendb/pull/21142/files#r2419026675

Scooletz commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -964,67 +967,58 @@ public static void WriteIncludedCounterNames(this AbstractBlittableJsonTextWrite
                 throw new NotSupportedException($"Cannot write query result of '{typeof(TResult)}' type in '{result.GetType()}'.");
 
+            (numberOfResults, totalDocumentsSizeInBytes) = writeResultsAsync.IsCompletedSuccessfully 
+                ? writeResultsAsync.Result 
```
Similar to the above. Isn't it implemented by the compiler? Are there any gains here?

maciejaszyk replied:
Answer here: https://github.com/ravendb/ravendb/pull/21142/files#r2419026675

Scooletz commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -964,67 +967,58 @@ public static void WriteIncludedCounterNames(this AbstractBlittableJsonTextWrite
                 throw new NotSupportedException($"Cannot write query includes of '{includes.GetType()}' type in '{result.GetType()}'.");
+            }
 
+            if (writeIncludesAsync.IsCompletedSuccessfully == false)
```
Similar to the above. Isn't it implemented by the compiler?

maciejaszyk replied:
Answer here: https://github.com/ravendb/ravendb/pull/21142/files#r2419026675

Scooletz reviewed (commented):
Provided some comments. Mostly, is the check `.IsCompletedSuccessfully` provides anything? The main question is that if we had the disposal scope, would it make it better anyway?

maciejaszyk committed:
RavenDB-24102 Read parameters for GET operation in sync manner.

maciejaszyk committed:
RavenDB-24102

- Moves usings outside execute task to generate better state machine
- Parse GET results in sync manner
- Removes duplicated reference

arekpalinski commented on bench/Requests.Benchmark/QueriesPostHandlerBenchmark.cs:
```diff
@@ -0,0 +1,62 @@
+[InvocationCount(1024)]
+public class QueriesPostHandlerBenchmark
+{
+    private string PathToRequests = @"D:\search.reqs";
```
What is `search.reqs`? Can you upload to a well known location and reference it in a comment here? This way another dev will be able to run it.

maciejaszyk replied:
QA file.

arekpalinski replied:
Please leave a comment from where to download it

arekpalinski commented on bench/Requests.Benchmark/QueriesPostHandlerBenchmark.cs:
```diff
@@ -0,0 +1,62 @@
+namespace Requests.Benchmark;
+
+[InvocationCount(1024)]
+public class QueriesPostHandlerBenchmark
```
Can you please show the results? How does it compare to previous impl?

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -78,145 +81,190 @@ private async ValueTask IndexEntriesAsync(TQueryContext queryContext, TOperation
+        try
         {
-            try
+            return await task;
```
Is `ExecuteWithExceptionHandling` truly necessary?

It may be clearer and simpler to use an explicit try/catch. This approach avoids the intermediate `await task;` call and keeps the flow straightforward. We can still extract the error handling code (content of `catch` block) into a private utility method and reuse it within the catch blocks as needed, without introducing a broader execution wrapper.

maciejaszyk replied:
I removed this one. However, ExecuteWithExceptionHandling(ValueTask) allows us to propagate the async Execute call from the processor and push the task to be awaited in upper levels (the QueryHandler or even the router).

redknightlois commented on bench/Requests.Benchmark/Program.cs:
```diff
@@ -0,0 +1,28 @@
+        Main(string[] args)
+    {
+#if DEBUG
+        var benchmark = new QueriesPostHandlerBenchmark()
```
What is the scenario this bench is covering? How can this be migrated into https://github.com/redknightlois/Raven.Bench with the objective of being included as a standard regression test?

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessor.cs:
```diff
@@ -36,24 +38,33 @@ protected AbstractQueriesHandlerProcessor([NotNull] TRequestHandler requestHandl
+        return ReadIndexQueryForPost(context, tracker, addSpatialProperties);
+    }
+    
+    internal async ValueTask<IndexQueryServerSide> ReadIndexQueryForPost(JsonOperationContext context, RequestTimeTracker tracker, bool addSpatialProperties)
```
```suggestion
    internal async ValueTask<IndexQueryServerSide> ReadIndexQueryForPostAsync(JsonOperationContext context, RequestTimeTracker tracker, bool addSpatialProperties)
```

Scooletz commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessor.cs:
```diff
@@ -36,24 +38,33 @@ protected AbstractQueriesHandlerProcessor([NotNull] TRequestHandler requestHandl
+    {
+        Debug.Assert(QueryMethod != HttpMethod.Get);
+        var readJsonTask = context.ReadForMemoryAsync(_stream, "index/query");
+        var json = readJsonTask.IsCompletedSuccessfully ? readJsonTask.Result : await readJsonTask;
```
I'd double check the generated IL. As far as I remember this check is emitted already. The question is how likely is it that it's `.IsCompletedSuccessfully ` and weather you want to delegate the `await readJsonTask` to a separate async method

Scooletz commented on src/Raven.Server/Documents/Handlers/QueriesHandler.cs:
```diff
@@ -11,14 +11,26 @@ public sealed class QueriesHandler : DatabaseRequestHandler
         {
             using (var processor = new DatabaseQueriesHandlerProcessorForGet(this, HttpMethod.Post))
-                await processor.ExecuteAsync();
+            using (processor.AllocateContextForQueryOperation(out var queryContext, out var context))
```
A lot of usings here. I know it's top lvl so we should be fine.

Scooletz commented:
Please reconsider stacking on top of https://github.com/ravendb/ravendb/pull/21571 once it's merged in. This can potentially help a lot with the scoping and `using` clauses.

maciejaszyk committed:
RavenDB-24102 Remove Requests.Benchmark project and related code.

maciejaszyk committed:
RavenDB-24102 Register disposables from processor into request context scope

maciejaszyk committed:
RavenDB-24102 Remove unused project references from solution file.

maciejaszyk committed:
RavenDB-24102 Simplify exception handling for query loading and remove unused generic method.

redknightlois commented:
<img width="1323" height="557" alt="image" src="https://github.com/user-attachments/assets/e896c647-1325-4afd-8f6c-235d0ce09bb8" />

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessor.cs:
```diff
@@ -38,24 +40,42 @@ protected AbstractQueriesHandlerProcessor([NotNull] TRequestHandler requestHandl
     protected abstract RavenConfiguration Configuration { get; }
 
-    protected RequestTimeTracker CreateRequestTimeTracker()
+    internal RequestTimeTracker CreateRequestTimeTracker()
```
Do we need it to be `internal`? I _think_ it can remain `protected`

arekpalinski replied:
@maciejaszyk ping

maciejaszyk committed:
RavenDB-24102 Add benchmark for RequestContext query scoping

maciejaszyk commented on bench/RequestHandler.Benchmark/RequestContextQueryScopingBenchmark.cs:
```diff
@@ -0,0 +1,144 @@
+namespace RequestHandler.Benchmark;
+
+[MemoryDiagnoser]
+public class RequestContextQueryScopingBenchmark
```
v6.2:
| Method            | Mean     | Error    | StdDev   | Gen0   | Gen1   | Allocated |
|------------------ |---------:|---------:|---------:|-------:|-------:|----------:|
| QueryGetDocument  | 18.94 us | 0.254 us | 0.237 us | 1.7090 | 0.7324 |  29.14 KB |
| QueryPostDocument | 12.30 us | 0.204 us | 0.181 us | 1.0986 | 0.4883 |  17.96 KB |

this PR:
| Method            | Mean     | Error    | StdDev   | Gen0   | Gen1   | Allocated |
|------------------ |---------:|---------:|---------:|-------:|-------:|----------:|
| QueryGetDocument  | 19.57 us | 0.381 us | 0.391 us | 1.7090 | 0.7324 |  28.92 KB |
| QueryPostDocument | 12.25 us | 0.169 us | 0.149 us | 0.9766 | 0.4883 |  17.93 KB |

ppekrol commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessorForGet.cs:
```diff
@@ -76,147 +84,194 @@ private async ValueTask IndexEntriesAsync(TQueryContext queryContext, TOperation
+    }
 
-                    EnsureQueryContextInitialized(queryContext, indexQuery);
+    public async ValueTask LoadIndexQueryForPost()
```
LoadIndexQueryForPostAsync - let's keep naming conventions

redknightlois commented:
After @maciejaszyk asked me to implement the index query and the voron query. I am not so convinced on this, we got higher CPU usage. Something is off on this codepath.
<img width="1323" height="600" alt="image" src="https://github.com/user-attachments/assets/fcd884b2-3f56-4d71-9cac-75f5eca31eef" />

redknightlois commented:
After poking around with @maciejaszyk I built some extensions to Raven.Bench for the scenario we wanted to run and this is the result on Lucene (instead of Corax). With these numbers the PR is a no-go. We will work on getting profiler runs to have a better idea of what is going on. 

<img width="1318" height="561" alt="image" src="https://github.com/user-attachments/assets/28930e99-8f9f-41af-94df-930e2a323a16" />

redknightlois commented:
Still slower throughput wise, but this clears up that cpu consumption is dominated by async machinery. 
<img width="1312" height="263" alt="image" src="https://github.com/user-attachments/assets/d19f03a5-e7e5-41d5-9ff8-286065897e7f" />

arekpalinski commented:
I talked to @maciejaszyk and we believe it's good idea to test in on v7.2 branch which is based on .NET10.

CC @redknightlois

maciejaszyk committed:
RavenDB-24102 Refactor query processing methods to improve exception handling and access modifiers

maciejaszyk commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -966,67 +969,58 @@ public static void WriteIncludedCounterNames(this AbstractBlittableJsonTextWrite
-                var maybeFlush = writer.MaybeFlushAsync(token);
-                if (maybeFlush.IsCompletedSuccessfully == false)    
-                    await maybeFlush;
+                writeResultsAsync = ValueTask.FromResult<(long NumberOfResults, long TotalDocumentsSizeInBytes)>((numberOfResults, totalDocumentsSizeInBytes));
             }
             else if (results is List<SuggestionResult> suggestions)
             {
                 numberOfResults = suggestions.Count;
                 writer.WriteArray(context, nameof(result.Results), suggestions, (w, c, suggestion) => w.WriteSuggestionResult(c, suggestion));
-                writer.WriteComma();
-                var maybeFlush = writer.MaybeFlushAsync(token);
-                if (maybeFlush.IsCompletedSuccessfully == false)    
-                    await maybeFlush;
+                writeResultsAsync = ValueTask.FromResult<(long NumberOfResults, long TotalDocumentsSizeInBytes)>((numberOfResults, totalDocumentsSizeInBytes));
```
Most of them flushes internally, however just for safety I've added it again.

maciejaszyk commented on src/Raven.Server/Documents/Handlers/Processors/Queries/DatabaseQueriesHandlerProcessorForGet.cs:
```diff
@@ -15,64 +16,75 @@
     {
+        Debug.Assert(QueryContext is not null);
+        Debug.Assert(IndexQuery is not null);
+        QueryContext.WithQuery(IndexQuery.Metadata);
```
Actually, this method shouldn't have any parameters anymore since context and query are fields now. It should assert them instead. Fixed.

maciejaszyk committed:
RavenDB-24102 Refactor query context initialization methods and enhance flushing logic in JSON writer

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Queries/AbstractQueriesHandlerProcessor.cs:
```diff
@@ -66,11 +86,13 @@ public async ValueTask<IndexQueryServerSide> GetIndexQueryAsync(JsonOperationCon
-
-    private async ValueTask<IndexQueryServerSide> ReadIndexQueryAsync(JsonOperationContext context, RequestTimeTracker tracker, bool addSpatialProperties)
+    
+    internal IndexQueryServerSide ReadIndexQueryForGet(JsonOperationContext context, RequestTimeTracker tracker, bool addSpatialProperties)
```
Can you also check if `internal` is needed here? Could it be `protected`?

maciejaszyk replied:
Modified. `CreateRequestTimeTracker` as well.

arekpalinski commented on src/Sparrow/Json/AsyncBlittableJsonTextWriter.cs:
```diff
@@ -38,6 +38,12 @@ public async ValueTask WriteStreamAsync(Stream stream, CancellationToken token =
+        public bool ShouldFlushAsync
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _innerStream.Length * 2 > _innerStream.Capacity;
```
This is new, right? Do we have any similar condition anywhere else? I wonder if this can result in some behavior changes hence affecting the system?

maciejaszyk replied:
```csharp
        public bool ShouldFlushAsync
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get => _innerStream.Length * 2 > _innerStream.Capacity;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public ValueTask<int> MaybeFlushAsync(CancellationToken token = default)
        {
            // PERF: Use cached MemoryStream reference
            if (_innerStream.Length * 2 <= _innerStream.Capacity)
                return new ValueTask<int>(0);

            FlushInternal(); // this is OK, because inner stream is a MemoryStream
            return FlushAsync(token);
        }
```


This is a shortcut to not to call `MaybeFlushAsync` each time.

maciejaszyk committed:
RavenDB-24102 Refactor access modifiers for query handling methods to enhance encapsulation
